Better server errors

**Current**
```
[INFO] File read: ./examples/jsonplaceholder_script.graphql ... ok
[INFO] File read: ./examples/scripts/echo.js ... ok
[INFO] N + 1: 1
Error: Server Failed
       ❯ The port is already in use
Caused by:
  • error creating server listener: Address already in use (os error 48)
```

**Improved**
```
[INFO] File read: ./examples/jsonplaceholder_script.graphql ... ok
[INFO] File read: ./examples/scripts/echo.js ... ok
[INFO] N + 1: 1
[ERROR] Server Failed: The port is already in use
[ERROR]   Caused by:
[ERROR]   • error creating server listener: Address already in use (os error 48)```
```

Use color coding
`[ERROR]` - RED
`Caused by:` — GRAY


**Technical Requirements**
- Make sure existing tests are updated and passing.

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -84,13 +84,7 @@ fn bullet(str: &str) -> String {
 
 impl Display for CLIError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let error_prefix = "[ERROR] ";
-        let colored_error_prefix = self.colored(error_prefix, colored::Color::Red);
         let default_padding = 2;
-
-        if self.is_root {
-            f.write_str(&colored_error_prefix)?;
-        }
 
         let message_color = if self.is_root {
             colored::Color::Yellow
@@ -98,11 +92,16 @@ impl Display for CLIError {
             colored::Color::White
         };
 
-        f.write_str(self.colored(&self.message, message_color).as_str())?;
-
         if let Some(description) = &self.description {
-            f.write_str(&self.colored(": ", message_color))?;
-            f.write_str(&self.colored(description.to_string().as_str(), colored::Color::White))?;
+            log::error!("{}: {}", self.colored(&self.message, message_color).as_str(), self.colored(description.to_string().as_str(), colored::Color::White));
+        }
+        else {
+            if self.is_root {
+                log::error!("{}", self.colored(&self.message, message_color).as_str());
+            }
+            else {
+                log::error!("{}", margin(bullet(self.message.as_str()).as_str(), default_padding));       
+            }
         }
 
         if !self.trace.is_empty() {
@@ -120,16 +119,14 @@ impl Display for CLIError {
         }
 
         if !self.caused_by.is_empty() {
-            f.write_str("\n")?;
             log::error!("{}", self.dimmed("Caused by:"));   
 
             for (i, error) in self.caused_by.iter().enumerate() {
                 let message = &error.to_string();
-                log::error!("{}", margin(bullet(message.as_str()).as_str(), default_padding));       
 
-                if i < self.caused_by.len() - 1 {
-                    f.write_str("\n")?;
-                }
+                if message.len() > 0 {
+                   log::error!("{}", margin(bullet(message.as_str()).as_str(), default_padding));   
+                }  
             }
         }
 

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -121,13 +121,11 @@ impl Display for CLIError {
 
         if !self.caused_by.is_empty() {
             f.write_str("\n")?;
-            f.write_str(&colored_error_prefix)?;
-            f.write_str(self.dimmed("Caused by:\n").as_str())?;
+            log::error!("{}", self.dimmed("Caused by:"));   
+
             for (i, error) in self.caused_by.iter().enumerate() {
                 let message = &error.to_string();
-                f.write_str(&colored_error_prefix)?;
-
-                f.write_str(&margin(bullet(message.as_str()).as_str(), default_padding))?;
+                log::error!("{}", margin(bullet(message.as_str()).as_str(), default_padding));       
 
                 if i < self.caused_by.len() - 1 {
                     f.write_str("\n")?;

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -93,15 +93,18 @@ impl Display for CLIError {
         };
 
         if let Some(description) = &self.description {
-            log::error!("{}: {}", self.colored(&self.message, message_color).as_str(), self.colored(description.to_string().as_str(), colored::Color::White));
-        }
-        else {
-            if self.is_root {
-                log::error!("{}", self.colored(&self.message, message_color).as_str());
-            }
-            else {
-                log::error!("{}", margin(bullet(self.message.as_str()).as_str(), default_padding));       
-            }
+            log::error!(
+                "{}: {}",
+                self.colored(&self.message, message_color).as_str(),
+                self.colored(description.to_string().as_str(), colored::Color::White)
+            );
+        } else if self.is_root {
+            log::error!("{}", self.colored(&self.message, message_color).as_str());
+        } else {
+            log::error!(
+                "{}",
+                margin(bullet(self.message.as_str()).as_str(), default_padding)
+            );
         }
 
         if !self.trace.is_empty() {
@@ -119,14 +122,17 @@ impl Display for CLIError {
         }
 
         if !self.caused_by.is_empty() {
-            log::error!("{}", self.dimmed("Caused by:"));   
+            log::error!("{}", self.dimmed("Caused by:"));
 
-            for (i, error) in self.caused_by.iter().enumerate() {
+            for error in self.caused_by.iter() {
                 let message = &error.to_string();
 
-                if message.len() > 0 {
-                   log::error!("{}", margin(bullet(message.as_str()).as_str(), default_padding));   
-                }  
+                if !message.is_empty() {
+                    log::error!(
+                        "{}",
+                        margin(bullet(message.as_str()).as_str(), default_padding)
+                    );
+                }
             }
         }
 

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -84,13 +84,7 @@ fn bullet(str: &str) -> String {
 
 impl Display for CLIError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let error_prefix = "[ERROR] ";
-        let colored_error_prefix = self.colored(error_prefix, colored::Color::Red);
         let default_padding = 2;
-
-        if self.is_root {
-            f.write_str(&colored_error_prefix)?;
-        }
 
         let message_color = if self.is_root {
             colored::Color::Yellow
@@ -121,11 +115,10 @@ impl Display for CLIError {
 
         if !self.caused_by.is_empty() {
             f.write_str("\n")?;
-            f.write_str(&colored_error_prefix)?;
-            f.write_str(self.dimmed("Caused by:\n").as_str())?;
+            f.write_str(self.dimmed("Caused by:").as_str())?;
+            f.write_str("\n")?;
             for (i, error) in self.caused_by.iter().enumerate() {
                 let message = &error.to_string();
-                f.write_str(&colored_error_prefix)?;
 
                 f.write_str(&margin(bullet(message.as_str()).as_str(), default_padding))?;
 
@@ -258,7 +251,7 @@ mod tests {
     #[test]
     fn test_title() {
         let error = CLIError::new("Server could not be started");
-        let expected = r"[ERROR] Server could not be started".strip_margin();
+        let expected = r"Server could not be started".strip_margin();
         assert_eq!(error.to_string(), expected);
     }
 
@@ -266,8 +259,7 @@ mod tests {
     fn test_title_description() {
         let error = CLIError::new("Server could not be started")
             .description("The port is already in use".to_string());
-        let expected =
-            r"|[ERROR] Server could not be started: The port is already in use".strip_margin();
+        let expected = r"|Server could not be started: The port is already in use".strip_margin();
 
         assert_eq!(error.to_string(), expected);
     }
@@ -279,7 +271,7 @@ mod tests {
             .trace(vec!["@server".into(), "port".into()]);
 
         let expected =
-            r"|[ERROR] Server could not be started: The port is already in use [at @server.port]"
+            r"|Server could not be started: The port is already in use [at @server.port]"
                 .strip_margin();
 
         assert_eq!(error.to_string(), expected);
@@ -297,9 +289,9 @@ mod tests {
             "baseURL".into(),
         ])]);
 
-        let expected = r"|[ERROR] Configuration Error
-                     |[ERROR] Caused by:
-                     |[ERROR]   • Base URL needs to be specified [at User.posts.@http.baseURL]"
+        let expected = r"|Configuration Error
+                     |Caused by:
+                     |  • Base URL needs to be specified [at User.posts.@http.baseURL]"
             .strip_margin();
 
         assert_eq!(error.to_string(), expected);
@@ -336,12 +328,12 @@ mod tests {
             ]),
         ]);
 
-        let expected = r"|[ERROR] Configuration Error
-                     |[ERROR] Caused by:
-                     |[ERROR]   • Base URL needs to be specified [at User.posts.@http.baseURL]
-                     |[ERROR]   • Base URL needs to be specified [at Post.users.@http.baseURL]
-                     |[ERROR]   • Base URL needs to be specified: Set `baseURL` in @http or @server directives [at Query.users.@http.baseURL]
-                     |[ERROR]   • Base URL needs to be specified [at Query.posts.@http.baseURL]"
+        let expected = r"|Configuration Error
+                     |Caused by:
+                     |  • Base URL needs to be specified [at User.posts.@http.baseURL]
+                     |  • Base URL needs to be specified [at Post.users.@http.baseURL]
+                     |  • Base URL needs to be specified: Set `baseURL` in @http or @server directives [at Query.users.@http.baseURL]
+                     |  • Base URL needs to be specified [at Query.posts.@http.baseURL]"
             .strip_margin();
 
         assert_eq!(error.to_string(), expected);
@@ -354,9 +346,9 @@ mod tests {
             .trace(vec!["Query", "users", "@http", "baseURL"]);
         let valid = ValidationError::from(cause);
         let error = CLIError::from(valid);
-        let expected = r"|[ERROR] Invalid Configuration
-                     |[ERROR] Caused by:
-                     |[ERROR]   • Base URL needs to be specified: Set `baseURL` in @http or @server directives [at Query.users.@http.baseURL]"
+        let expected = r"|Invalid Configuration
+                     |Caused by:
+                     |  • Base URL needs to be specified: Set `baseURL` in @http or @server directives [at Query.users.@http.baseURL]"
             .strip_margin();
 
         assert_eq!(error.to_string(), expected);

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -84,7 +84,13 @@ fn bullet(str: &str) -> String {
 
 impl Display for CLIError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let error_prefix = "[ERROR] ";
+        let colored_error_prefix = self.colored(error_prefix, colored::Color::Red);
         let default_padding = 2;
+
+        if self.is_root {
+            f.write_str(&colored_error_prefix)?;
+        }
 
         let message_color = if self.is_root {
             colored::Color::Yellow
@@ -92,19 +98,11 @@ impl Display for CLIError {
             colored::Color::White
         };
 
+        f.write_str(self.colored(&self.message, message_color).as_str())?;
+
         if let Some(description) = &self.description {
-            log::error!(
-                "{}: {}",
-                self.colored(&self.message, message_color).as_str(),
-                self.colored(description.to_string().as_str(), colored::Color::White)
-            );
-        } else if self.is_root {
-            log::error!("{}", self.colored(&self.message, message_color).as_str());
-        } else {
-            log::error!(
-                "{}",
-                margin(bullet(self.message.as_str()).as_str(), default_padding)
-            );
+            f.write_str(&self.colored(": ", message_color))?;
+            f.write_str(&self.colored(description.to_string().as_str(), colored::Color::White))?;
         }
 
         if !self.trace.is_empty() {
@@ -122,16 +120,17 @@ impl Display for CLIError {
         }
 
         if !self.caused_by.is_empty() {
-            log::error!("{}", self.dimmed("Caused by:"));
-
-            for error in self.caused_by.iter() {
+            f.write_str("\n")?;
+            f.write_str(&colored_error_prefix)?;
+            f.write_str(self.dimmed("Caused by:\n").as_str())?;
+            for (i, error) in self.caused_by.iter().enumerate() {
                 let message = &error.to_string();
+                f.write_str(&colored_error_prefix)?;
 
-                if !message.is_empty() {
-                    log::error!(
-                        "{}",
-                        margin(bullet(message.as_str()).as_str(), default_padding)
-                    );
+                f.write_str(&margin(bullet(message.as_str()).as_str(), default_padding))?;
+
+                if i < self.caused_by.len() - 1 {
+                    f.write_str("\n")?;
                 }
             }
         }

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -85,18 +85,24 @@ fn bullet(str: &str) -> String {
 impl Display for CLIError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         let error_prefix = "[ERROR] ";
+        let colored_error_prefix = self.colored(error_prefix, colored::Color::Red);
         let default_padding = 2;
 
         if self.is_root {
-            f.write_str(self.colored(error_prefix, colored::Color::Red).as_str())?;
+            f.write_str(&colored_error_prefix)?;
         }
 
-        f.write_str(&self.message.to_string())?;
+        let message_color = if self.is_root {
+            colored::Color::Yellow
+        } else {
+            colored::Color::White
+        };
+
+        f.write_str(self.colored(&self.message, message_color).as_str())?;
 
         if let Some(description) = &self.description {
-            f.write_str(
-                &self.colored(format!(": {}", description).as_str(), colored::Color::White),
-            )?;
+            f.write_str(&self.colored(": ", message_color))?;
+            f.write_str(&self.colored(description.to_string().as_str(), colored::Color::White))?;
         }
 
         if !self.trace.is_empty() {
@@ -115,11 +121,11 @@ impl Display for CLIError {
 
         if !self.caused_by.is_empty() {
             f.write_str("\n")?;
-            f.write_str(self.colored(error_prefix, colored::Color::Red).as_str())?;
+            f.write_str(&colored_error_prefix)?;
             f.write_str(self.dimmed("Caused by:\n").as_str())?;
             for (i, error) in self.caused_by.iter().enumerate() {
                 let message = &error.to_string();
-                f.write_str(self.colored(error_prefix, colored::Color::Red).as_str())?;
+                f.write_str(&colored_error_prefix)?;
 
                 f.write_str(&margin(bullet(message.as_str()).as_str(), default_padding))?;
 

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -260,8 +260,8 @@ mod tests {
     fn test_title_description() {
         let error = CLIError::new("Server could not be started")
             .description("The port is already in use".to_string());
-        let expected = r"|[ERROR] Server could not be started: The port is already in use"
-            .strip_margin();
+        let expected =
+            r"|[ERROR] Server could not be started: The port is already in use".strip_margin();
 
         assert_eq!(error.to_string(), expected);
     }
@@ -272,8 +272,9 @@ mod tests {
             .description("The port is already in use".to_string())
             .trace(vec!["@server".into(), "port".into()]);
 
-        let expected = r"|[ERROR] Server could not be started: The port is already in use [at @server.port]"
-            .strip_margin();
+        let expected =
+            r"|[ERROR] Server could not be started: The port is already in use [at @server.port]"
+                .strip_margin();
 
         assert_eq!(error.to_string(), expected);
     }

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -84,13 +84,8 @@ fn bullet(str: &str) -> String {
 
 impl Display for CLIError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let error_prefix = "Error: ";
+        let error_prefix = "[ERROR] ";
         let default_padding = 2;
-        let root_padding_size = if self.is_root {
-            error_prefix.len()
-        } else {
-            default_padding
-        };
 
         if self.is_root {
             f.write_str(self.colored(error_prefix, colored::Color::Red).as_str())?;
@@ -99,18 +94,8 @@ impl Display for CLIError {
         f.write_str(&self.message.to_string())?;
 
         if let Some(description) = &self.description {
-            f.write_str("\n")?;
-            let color = if self.is_root {
-                colored::Color::Yellow
-            } else {
-                colored::Color::White
-            };
             f.write_str(
-                margin(
-                    &self.colored(format!("❯ {}", description).as_str(), color),
-                    root_padding_size,
-                )
-                .as_str(),
+                &self.colored(format!(": {}", description).as_str(), colored::Color::White),
             )?;
         }
 
@@ -129,9 +114,13 @@ impl Display for CLIError {
         }
 
         if !self.caused_by.is_empty() {
-            f.write_str(self.dimmed("\nCaused by:\n").as_str())?;
+            f.write_str("\n")?;
+            f.write_str(self.colored(error_prefix, colored::Color::Red).as_str())?;
+            f.write_str(self.dimmed("Caused by:\n").as_str())?;
             for (i, error) in self.caused_by.iter().enumerate() {
                 let message = &error.to_string();
+                f.write_str(self.colored(error_prefix, colored::Color::Red).as_str())?;
+
                 f.write_str(&margin(bullet(message.as_str()).as_str(), default_padding))?;
 
                 if i < self.caused_by.len() - 1 {

--- a/src/cli/error.rs
+++ b/src/cli/error.rs
@@ -252,7 +252,7 @@ mod tests {
     #[test]
     fn test_title() {
         let error = CLIError::new("Server could not be started");
-        let expected = r"Error: Server could not be started".strip_margin();
+        let expected = r"[ERROR] Server could not be started".strip_margin();
         assert_eq!(error.to_string(), expected);
     }
 
@@ -260,8 +260,7 @@ mod tests {
     fn test_title_description() {
         let error = CLIError::new("Server could not be started")
             .description("The port is already in use".to_string());
-        let expected = r"|Error: Server could not be started
-                     |       ❯ The port is already in use"
+        let expected = r"|[ERROR] Server could not be started: The port is already in use"
             .strip_margin();
 
         assert_eq!(error.to_string(), expected);
@@ -273,8 +272,7 @@ mod tests {
             .description("The port is already in use".to_string())
             .trace(vec!["@server".into(), "port".into()]);
 
-        let expected = r"|Error: Server could not be started
-                     |       ❯ The port is already in use [at @server.port]"
+        let expected = r"|[ERROR] Server could not be started: The port is already in use [at @server.port]"
             .strip_margin();
 
         assert_eq!(error.to_string(), expected);
@@ -292,9 +290,9 @@ mod tests {
             "baseURL".into(),
         ])]);
 
-        let expected = r"|Error: Configuration Error
-                     |Caused by:
-                     |  • Base URL needs to be specified [at User.posts.@http.baseURL]"
+        let expected = r"|[ERROR] Configuration Error
+                     |[ERROR] Caused by:
+                     |[ERROR]   • Base URL needs to be specified [at User.posts.@http.baseURL]"
             .strip_margin();
 
         assert_eq!(error.to_string(), expected);
@@ -331,13 +329,12 @@ mod tests {
             ]),
         ]);
 
-        let expected = r"|Error: Configuration Error
-                     |Caused by:
-                     |  • Base URL needs to be specified [at User.posts.@http.baseURL]
-                     |  • Base URL needs to be specified [at Post.users.@http.baseURL]
-                     |  • Base URL needs to be specified
-                     |      ❯ Set `baseURL` in @http or @server directives [at Query.users.@http.baseURL]
-                     |  • Base URL needs to be specified [at Query.posts.@http.baseURL]"
+        let expected = r"|[ERROR] Configuration Error
+                     |[ERROR] Caused by:
+                     |[ERROR]   • Base URL needs to be specified [at User.posts.@http.baseURL]
+                     |[ERROR]   • Base URL needs to be specified [at Post.users.@http.baseURL]
+                     |[ERROR]   • Base URL needs to be specified: Set `baseURL` in @http or @server directives [at Query.users.@http.baseURL]
+                     |[ERROR]   • Base URL needs to be specified [at Query.posts.@http.baseURL]"
             .strip_margin();
 
         assert_eq!(error.to_string(), expected);
@@ -350,10 +347,9 @@ mod tests {
             .trace(vec!["Query", "users", "@http", "baseURL"]);
         let valid = ValidationError::from(cause);
         let error = CLIError::from(valid);
-        let expected = r"|Error: Invalid Configuration
-                     |Caused by:
-                     |  • Base URL needs to be specified
-                     |      ❯ Set `baseURL` in @http or @server directives [at Query.users.@http.baseURL]"
+        let expected = r"|[ERROR] Invalid Configuration
+                     |[ERROR] Caused by:
+                     |[ERROR]   • Base URL needs to be specified: Set `baseURL` in @http or @server directives [at Query.users.@http.baseURL]"
             .strip_margin();
 
         assert_eq!(error.to_string(), expected);

--- a/src/cli/tc.rs
+++ b/src/cli/tc.rs
@@ -194,12 +194,15 @@ fn logger_init() {
             let level = record.level();
             let color_styles = buf.default_level_style(level);
 
-            writeln!(
-                buf,
-                "{color_styles}[{}]{color_styles:#} {}",
-                record.level(),
-                record.args(),
-            )?;
+            for line in record.args().to_string().lines() {
+                writeln!(
+                    buf,
+                    "{color_styles}[{}]{color_styles:#} {}",
+                    record.level(),
+                    line,
+                )?;
+            }
+
             Ok(())
         })
         .init();

--- a/src/cli/tc.rs
+++ b/src/cli/tc.rs
@@ -193,18 +193,12 @@ fn logger_init() {
         .format(|buf, record| {
             let level = record.level();
             let color_styles = buf.default_level_style(level);
-            let mut lines_buffer = Vec::new();
-
-            for line in record.args().to_string().lines() {
-                writeln!(
-                    &mut lines_buffer,
-                    "{color_styles}[{}]{color_styles:#} {}",
-                    record.level(),
-                    line,
-                )?;
-            }
-
-            buf.write_all(&lines_buffer)?;
+            writeln!(
+                buf,
+                "{color_styles}[{}]{color_styles:#} {}",
+                record.level(),
+                record.args(),
+            )?;
 
             Ok(())
         })

--- a/src/cli/tc.rs
+++ b/src/cli/tc.rs
@@ -193,15 +193,18 @@ fn logger_init() {
         .format(|buf, record| {
             let level = record.level();
             let color_styles = buf.default_level_style(level);
+            let mut lines_buffer = Vec::new();
 
             for line in record.args().to_string().lines() {
                 writeln!(
-                    buf,
+                    &mut lines_buffer,
                     "{color_styles}[{}]{color_styles:#} {}",
                     record.level(),
                     line,
                 )?;
             }
+
+            buf.write_all(&lines_buffer)?;
 
             Ok(())
         })

--- a/src/main.rs
+++ b/src/main.rs
@@ -29,7 +29,7 @@ fn main() -> anyhow::Result<()> {
                     CLIError::new(&error.to_string()).caused_by(sources)
                 }
             };
-            eprintln!("{}", cli_error.color(true));
+            log::error!("{}", cli_error.color(true));
             std::process::exit(exitcode::CONFIG);
         }
     }

